README.md
# RSA Cryptography

## Usage:
```console
$ make
$ ./main
```

## Functions:

+ `generate_n_bit_long_prime`
    * Implement: The first thing to do is generate random numbers. Since the pseudo-random numbers generated using the same seeds are basically the same as ordinary numbers, the current time can be used as seeds as random numbers. After generating the random number, we need to ensure that the random number satisfies the provided parameters (specified bits). Then you need to make sure that the generated numbers are prime numbers. At first I tried to generate another one directly when a number was not random, but later found that it would be faster to directly increment the original number. The last step is to return the generated random number.
    * Function: This function is considered a helper function. It is an integral step in generating public key pairs and private key pairs.
+ `find_coprime`
    * Implement: From the point of view of mathematical principles, if you want to find numbers that are relatively prime to φ, try one by one slowly. But it can be optimized here. 65537 is suitable for most situations. So I start directly with 65537. If I encounter a number that is not relatively prime to 65537, I will slowly increment65537 until the condition is met again.
    * Function: This function can be understood as a helper function. When we generate private key pair and public key pair, there is a step to calculate φ = (p-1)*(q-1). Then we need to find a number that is relatively prime to φ. This function can be used at that time. Of course, this can actually be implemented directly in that function, but hierarchy can reduce the complexity of the code, so I think this can make the program easier to debug.
+ `generate_key_pair`
    * Implement: The first thing to do is to generate two large prime numbers using `generate_key_pair`. Next, subtract one from the two large prime numbers and multiply them together to get φ. Then multiply the original two large prime numbers directly to get n. The next step is to find numbers that are relatively prime to φ. The previous `find_coprime` can be used in this step. All that needs to be done is to find a number that is relatively prime to φ and define it as e. The next step is to find the multiplicative inverse of e mod φ. Since the entire program is implemented based on the GMP library, mpz_invert can better implement this function. After this step, all the most basic calculations to generate a key pair are done.
    * Function: What this function does is generate a private key pair and a public key pair. The private key pair is used to decrypt information, and the public key pair is used to encrypt information.
+ `encrypt`
    * Implement: Encryption is generally achieved by $c = m^e\mod n$. Since this program is implemented based on the GMP library, mpz_pown can fulfill this mission very well.
    * Function: This function is used to encrypt the numbers converted from text.
+ `string_to_number`
    * Implement: The string_to_number function converts a string to a number by interpreting each character in the string as a number in the base-256 number system. For example, the string $1234$ will be converted to $1 \times 256^3 + 2 \times 256^2 + 3 \times 256^1 + 4 \times 256^0$.
    * Function: Since RSA encryption is basically implemented based on mathematical calculations, it may be difficult to directly encrypt strings. So a better way is to convert the string to a 256-base number and then implement encryption.
+ `decrypt`
    * Implement: Decryption is generally achieved by $m = c^d\mod n$. Since this program is implemented based on the GMP library, mpz_pown can fulfill this mission very well.
    * Function: This function is used to decrypt numbers that were ever converted from strings.
+ `number_to_string`
    * Implement: The way this function implements converting numbers to strings is the opposite of converting strings to numbers before. This function will only operate a byte at the same time and convert it to the corresponding character.
    * Function: Since the string was converted to a number earlier, the number needs to be converted to a string now.

## References[^footnote]:

+ [RSA Algorithm](https://en.wikipedia.org/wiki/RSA_(cryptosystem)#Key_generation)
+ [Miller-Rabin Primality Test](https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test)
+ [Modular Multiplicative Inverse](https://en.wikipedia.org/wiki/Modular_multiplicative_inverse)
+ [Modular Exponentiation](https://en.wikipedia.org/wiki/Modular_exponentiation)
+ [Greatest Common Divisor](https://en.wikipedia.org/wiki/Greatest_common_divisor)
+ [Euclidean Algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm)
+ [Extended Euclidean Algorithm](https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm)
+ [Modular Exponentiation](https://en.wikipedia.org/wiki/Modular_exponentiation#Right-to-left_binary_method)

[^footnote]: When I first tried to write this project, I used conventional data types (such as int, long, long long). But then I found that using these conventional data types to represent very large prime numbers of 4096bits would be very difficult, and The subsequent operations on these two prime numbers will also become very difficult, so I switched to GMP Library. Although these mathematical theories are no longer necessary to complete the project, I still keep them.

## Shortage:

Since strings and numbers need to be converted to each other before encryption and decryption, and there are many conversion methods, this program may not be able to decrypt numbers generated by encryption using other conversion methods.

CHANGELOG
CHANGELOG
=========

Version 1.0.0.0 (2022-12-12)
--------------------------
- Initial release
- Added new function `is_prime`

Version 1.0.0.1 (2022-12-16)
--------------------------
- Replace Rabin-Miller Algo.txt with Rabin-Miller Algo.jpg (the flowchart should explain the Rabin-Miller algorithm better than text)
- Added new function `gcd`
- Added new function `mod_inverse`
- Added new function `eea` (extended Euclidean algorithm)
- Added new function `mod_exp`

Version 1.0.0.1 (2022-12-19)
--------------------------
- Added new function `generate_rsa_keys`
- Optimized the `generate_prime` function
```cpp
long long generate_prime(long long min, long long max) {
    while (true) {
        long long n = rand() % (max - min + 1) + min;   // rand() is replaced by Mersenne Twister
        if (is_prime(n)) return n;
    }
}
```
- Added new function `generate_4096bit_prime`

Version 1.0.0.2 (2022-12-21)
--------------------------
- Added new function `generate_n_bit_long_prime`
    + Since the Seed remains unchanged, the generated pseudo-random numbers have no meaning (they are all the same number)
    + The time() function is used but it takes one second to generate the next random number
    + After shortening the wait time to generate the next random number, I found that sometimes the function generates the same number. However, if this happens later, you can force another different random number to be generated. So this flaw can be improved later. And after the test, it is found that if the bit number of the generated random number is large (such as 4096 and above, the probability of generating the same random number will become smaller) [but the same last two random numbers may be a bug]

Version 1.0.0.3 (2022-12-23)
--------------------------
- Added new function `find_coprime`
- All mpz_t in `generate_n_bit_long_prime` have been replaced with mpz_class (the last time I used C to write this project was mpz_t, but C++ may be more suitable to use mpz_class...)
- Added new function `generate_key_pair`

Version 1.0.0.4 (2022-12-26)
--------------------------
- Added new function `number_to_string`
- Added new function `string_to_number`
- Added new function `encrypt`
- Added new function `decrypt`
- The main function is all completed
- When decrypting, the standard input does not seem to be useful and may even make mistakes, so it was removed
- Now part of main.cc is put into functions.cc and functions.hh

Version 1.0.0.5 (2022-12-26)
--------------------------
- Improved README.md

main.cc
#include <gmp.h>    
#include <gmpxx.h>  
#include <iostream>
#include <exception>
#include <fstream>

using namespace std;

#include "functions.hh"

int main() {
    cout << "Do you want to encrypt or decrypt? (e/d): ";
    char choice;
    cin >> choice;
    while (choice != 'e' && choice != 'd') {
        cout << "Invalid choice. Please enter e or d: ";
        cin >> choice;
    }
    if (choice == 'e') {
        cout << "Do you want to use a file or standard input? (f/s): ";
        cin >> choice;
        while (choice != 'f' && choice != 's') {
            cout << "Invalid choice. Please enter f or s: ";
            cin >> choice;
        }
        if (choice == 'f') {
            cout << "Enter file name: ";
            string file_name;
            cin >> file_name;
            ifstream file(file_name);
            while (!file.is_open()) {
                cout << "Something is wrong. Please enter a valid file name: ";
                cin >> file_name;
                file.open(file_name);
            }
            string contents;
            string line;
            while (getline(file, line)) {
                contents += line + "\n";
            }
            mpz_class m;
            string_to_number(contents, m);
            cout << "Do you already have a key pair? (y/n): ";
            while (choice != 'y' && choice != 'n') {
                cout << "Invalid choice. Please enter y or n: ";
                cin >> choice;
            }
            if (choice == 'y') {
                cout << "Enter the public key(The e in [n, e]): ";
                mpz_class e;
                cin >> e;
                cout << "Enter the modulus(The n in [n, e]): ";
                mpz_class n;
                cin >> n;
                mpz_class c;
                encrypt(m, e, n, c);
                cout << "Encrypted message: " << c << endl;
                file.close();
            } else if (choice == 'n') {
                mpz_class e, d, n;
                generate_key_pair(4096, e, d, n);
                mpz_class c;
                encrypt(m, e, n, c);
                ofstream output_file("output.txt");
                output_file << "Encrypted message(c): " << c << endl;
                output_file << "Part of private key(d): " << d << endl;
                output_file << "Modulus(n): " << n << endl;
                output_file << "Part of public key(e): " << e << endl;
                output_file.close();
                file.close();
            } else {
                throw bad_exception();
            }
        } else if (choice == 's') {
            cout << "Enter the message: ";
            string message;
            cin.ignore();
            getline(cin, message);
            mpz_class m;
            string_to_number(message, m);
            cout << "Do you already have a key pair? (y/n): ";
            cin >> choice;
            while (choice != 'y' && choice != 'n') {
                cout << "Invalid choice. Please enter y or n: ";
                cin >> choice;
            }
            if (choice == 'y') {
                cout << "Enter the public key(The e in [n, e]): ";
                mpz_class e;
                cin >> e;
                cout << "Enter the modulus(The n in [n, e]): ";
                mpz_class n;
                cin >> n;
                mpz_class c;
                encrypt(m, e, n, c);
                cout << "Encrypted message: " << c << endl;
            } else if (choice == 'n') {
                mpz_class e, d, n;
                generate_key_pair(4096, e, d, n);
                mpz_class c;
                encrypt(m, e, n, c);
                ofstream output_file("output.txt");
                output_file << "Encrypted message(c): " << c << endl;
                output_file << "Part of private key(d): " << d << endl;
                output_file << "Modulus(n): " << n << endl;
                output_file << "Part of public key(e): " << e << endl;
                output_file.close();
            } else {
                throw bad_exception();
            }
        } else {
            throw bad_exception();
        }
    } else if (choice == 'd') {
        cout << "**************************************************";
        cout << "ATTENTION: The file should be in the following format (this is line is not contained in input file):" << endl;
        cout << "Encrypted message(c): " << endl;
        cout << "Part of private key(d): " << endl;
        cout << "Modulus(n): " << endl;
        cout << "**************************************************" << endl;
        cout << "Enter file name: ";
        string file_name;
        cin >> file_name;
        ifstream file(file_name);
        while (!file.is_open()) {
            cout << "Something is wrong. Please enter a valid file name: ";
            cin >> file_name;
            file.open(file_name);
        }
        string line1, line2, line3;
        getline(file, line1);
        getline(file, line2);
        getline(file, line3);
        line1.erase(std::remove_if(line1.begin(), line1.end(), [](char c) { return !std::isdigit(c); }), line1.end());
        line2.erase(std::remove_if(line2.begin(), line2.end(), [](char c) { return !std::isdigit(c); }), line2.end());
        line3.erase(std::remove_if(line3.begin(), line3.end(), [](char c) { return !std::isdigit(c); }), line3.end());
        mpz_class c(line1);
        mpz_class d(line2);
        mpz_class n(line3);
        cout << "c: " << c << endl;
        cout << "d: " << d << endl;
        cout << "n: " << n << endl;
        mpz_class m;
        decrypt(c, d, n, m);
        string message;
        number_to_string(m, message);
        cout << "Decrypted message: " << message << endl;
        file.close();
    } else {
        throw bad_exception();
    }
}


input.cs
// Encrypted message(c): 342413604973373211802411742332102367183772809735544984927483561273012346182418822137991123367028874393648810528671135917202242797248714556517157443239532186219539702092380719992059296902124273750360605326755424815147688606653484542876181455355856801183964869257392684708253043480574904604802862780221597862858141636664356067848415301857645704926108163749539100230860179423422575804562910764693430099999922737368145028906199073327926695223449183470929443799633221962745005440960344033064408283346644431796597942307956562891125354981788772636458500239603925186232500935857484949119887615861387873811448433171308597511986086202004553440016879229961837274230246577378187889671947966145618086179356932791936075288392633899645451453886162533887064617299113614622659774169199338745234589612188296512766694177337392629295461157707183299214392970032940478588828633406770632327936805065030158627836387019746578551838519964583799184312578286083383561324197973098016527970154189194797032797235546670236181794785518793005350443008743086740034093092879160167621009162958934436248117503914355243394477449303687393223689534849310822945700576991983918380545767621530062088232976680490068133002230690053931019747191118950667213848110169379940821788663418095656411036532141007887815797557839938794961632851721093349341139876052998193925956007257121162733044618789783848994066445178234370387186039177779725025814612302950475978057080551138152497634238612097555291367782411490702720547334426200853729047441289265363035080487033653172703858279098555154162017236628658211140589426179148585240239180674144484098019556335782474933114562668710051795175761782971258825618056867416342552603030247572604330437318125217908215768011946004555672477015977166863371196861630956463822312035728356505085607416817161628102124906965716594237153798289975034332712828445874771606014843485577903513318989226319951967160580029304059114234851492283108843151265029166241124506754674011549818739705665432200980773053144887579978531868478228020976190726796446338540345033170206482971850427860125407293944571176807954262045290565469025348143864443597189642581135019485719774996449156094437813359566251019619435643098940596466154236821432653351182290530720971490874626687520308539784962842069260262202386163818565515533399876205678615963377487235917991048950059964620756941897748918419512995029306015271534046925909429770388375296397186179307375540274337121909394702149902029729427331976655423418819194569116480915 message(c): 136962640183528630745756109997692030306724930784830872681956639696267346817665580713859535441491148981471739661044871226427410274074426273211811468028826319067477613705954777422728139543827365965360004585745265187253594604396644222095377691194222166755671556829591567221591409066552245066377241405139832631623192032301119884010854886413049176005693733517698530198318741370417091782618353785975687506924088080924944913392217786906357908766929468959345764478155002542248140361134298878830228969760225127836262579258483420996032940673155353692074940821619754286146302326252227732317345238882216365551503743002449623131331104967392638339655033408942271915802840347384944853759513033376985633981327871876116416796867814442614299224592333664289314878625313708012065556976235971093113209214880410357719586066295814933471107720935080046306227467620206029221939728187847938984696630807523544167113072679271288634044708083301753038598866593774830156128996488895443872963709986150415213587853120413201742934873443514408875959764110413309054904057793427279722764898573332481183363001850776776775709996707914720300495904517575928692891504723186117694660081654092933736975074573199616585540486006406264623952220223383688689639153942196938920346031270209723757425645535441117187800511389899953432286631972731174051986217808662460411361173061483579446621192315916325534170701542035838051583301906770045245497110535722696310295173655594528026175774176818367865474397870642181713837224938908809075510255808425911062455300858501742960509499244129364507082811458981424173384538396009473795548510008662763211665868973194639230286743526317117776290579143960803543088430097997486995028767896947177880158909335602557452377348249758764521451301921094717831416630741224362792327697778465090495055757123219788198321051448706647939351168733221241820158586560118231714631524723630126739302797329624203450569955982982625446727860353089589263555167940376311635341083033702909417600836442175676427624010174125757080268070271845674463835927547296290318733773784138480321747903520153328540224259985150962679945654057106851067937647280277275285659537027862313856961977343108691959436796190991832844687688189217218227622611822648536670277873362878456360842799698061734611028470286804847447258397322470922883538291121380622417950220095435343852351453819049231313832159461827746251025976777923333921842414881907962239892813159784441220013283163860644541005946009392236070847493191405651938334391813065061
// Part of private key(d): 425176172147684349411342653580238183444929020195567331037610401645767463476663081647193126137401338598747091632633874221337126965838243785813584659138738485147460669917813252545406638846165935632353884404722402053038722185758255183962821856612086592417378997361130731344510496011283806272127870628862534244813816319335847281332081841872898584887763872281937496317033698123808876369003421150074776266439917651952255338007471750924387440472833262187718027623544357891363171988349874988477139690740193941283353551678930345367888399947857597342892676199172244698877642957910252370386348654161466683679147278095945820964560625428740609154769159820508618534394155730416039471965965967163531648840973097293149483349921779006572082460635318288843447015456205649638104370681311585475510734337238037249344886750341493909644955376068995425006663816900253634212145395726905321522837183342530437127647130796898927613068720089520269870774355320883945360591172500145787282018202434308327314634488653624766792379421580047795744062890468581629469109722944103856196240424977808501873442493098098745816860915235417801866153342016245125420831291603218703409854579500194244354461397782163073651041789724853400162472487404559514651970831589973870354826439907019607949784418432304903576552890209772648529163954769183975760535686480210089450185262738065721910423428979543302809690904372735754971287953122148165780765126490793629108133458625594443873307653460045151148353214727992083124531860466719161251389339789600539652470347393882214712921145749914578159301064411855213526830070186184277003005589100835030954323895178655134539675604502719284206023262246662274041476898336876094937095868984760793461770817867683182743080340024948475599002779064210393606159649539727921436979481856272523337920645238856678017782923349362842386367461249827991011509137202219410133298736543407846743198386052699771047217779101022934471480636273827583889217669855239821353770331042873548749730515709722793056370416183019625690775480411762352353830098459196737542334702108791724105303484059827176742331098648010932479346447822584321843104717300985492260298012229507373200352524952034343052610652326824217239804474967073794731680976469031120801171808645486793165517717956133902944391269763066563677483667562571885959850570052004274086249069899256601042629611702162611289516056451671239580331502368864812659896190154609356902117754959738237992096042298501531053812530368030904900336300795253822782148841064884333
// Modulus(n): 581303239679624266347578251542465209730474876323289791889264063683293256573903585728843139828243903781080382691726842909017863564413079857950660223343600794849465524656383167457355061855871011234746563517936623831229763969709789715059381579572052133248310573694720449361117854951319637251620825230077477976475708357574077882062389645787507135867098798325677223242504171773027272986239224197610318393108853304599873956127999919481205375743570950328496430090001597645275231096286341026803511086096591015539525226168395912055393805515442648525318813394495700444984751987745159269803069401226244749103583501920809455899747777380272729783584154107784986604539267426812892017820663640137610799417771020700972936065585138849561167585744379987398205633711345565042921584256620786008314321398968796228440927150456461590829277990642197833913877679528359704294594196302371838044063429325532851946064671138758006028532068603460726536370896519552959780892117912632824889989087992850002028146458389331518541194746053856104927060595653268681558715842549029611422405522723910207307391272977367216117651273636874444161929520720113837169187866022742133417578795800651511260213552236353840781648602757853703691414611599720807588321944089123136360577041161914180325389756551644168234525393774612872045176712781888132445845598583579491721759639334567128001583556463899049496574575380498211013110043637299066553667465113096286369294305378620775340633050481760979064351902818356942001397442109584641039892648016199776456944363343150625410569499270879084464965791315370211673682093317709493037309460309097016256151154208109441910720777372958706367583216035002493837817913187108766126023749286276384825036094702488153641861915119518451751626966714949547983716247735022312602767052447562380127015157773120193479642545387125996481200441203776350206358489814472648043771787857724658885453819782623459379925502639201636480230853557575565682984028428805810073943689378274641680009980506388701260582751438856006920212383275694328848254475073867021131532091894177031604561155789740866963061741406164997283819077886986505017388352322812383934959148421812221847077113597299486979154890143748388749504328389771473773758879113337340748818793169939400342651743680515592193367833076288615022564421407816598156707945852287085681086290939761072238950811707130893184971299248287279513893207630943200984968161025341799946855898030749942559633849214152291285675358354920410627927390864631471350456887738798921


functions.cc
#include "functions.hh"

void generate_n_bit_long_prime(mpz_class &n, int num_bits) {
    gmp_randstate_t state;
    gmp_randinit_default(state);

    clock_t seed = clock();
    seed = seed / (CLOCKS_PER_SEC / 1000000);
    gmp_randseed_ui(state, seed);
    mpz_urandomb(n.get_mpz_t(), state, num_bits);

    mpz_setbit(n.get_mpz_t(), 0);
    mpz_setbit(n.get_mpz_t(), num_bits - 1);

    while (mpz_probab_prime_p(n.get_mpz_t(), 25) == 0) {
        mpz_add_ui(n.get_mpz_t(), n.get_mpz_t(), 2);
    }

    gmp_randclear(state);
}

mpz_class find_coprime(mpz_class &phi) {
    mpz_class e;
    e = 65537;

    while (mpz_gcd_ui(NULL, e.get_mpz_t(), phi.get_ui()) != 1) {
        mpz_add_ui(e.get_mpz_t(), e.get_mpz_t(), 2);
    }
    return e;
}

void generate_key_pair(int num_bits, mpz_class &e, mpz_class &d, mpz_class &n) {
    mpz_class p, q, phi;
    generate_n_bit_long_prime(p, num_bits);
    generate_n_bit_long_prime(q, num_bits);
    phi = (p - 1) * (q - 1);
    n = p * q;
    e = find_coprime(phi);
    mpz_invert(d.get_mpz_t(), e.get_mpz_t(), phi.get_mpz_t());
}

void encrypt(mpz_class &m, mpz_class &e, mpz_class &n, mpz_class &c) {
    mpz_powm(c.get_mpz_t(), m.get_mpz_t(), e.get_mpz_t(), n.get_mpz_t());
}

void string_to_number(const std::string& str, mpz_class& n) {
    n = 0;
    for (char c : str) {
        n <<= 8;
        n += c;
    }
}

void number_to_string(const mpz_class& n, std::string& str) {
    str = "";
    mpz_class m = n;
    while (m != 0) {
        char c = m.get_si() & 0xff;
        str = c + str;
        m >>= 8;
    }
}

void decrypt(mpz_class &c, mpz_class &d, mpz_class &n, mpz_class &m) {
    mpz_powm(m.get_mpz_t(), c.get_mpz_t(), d.get_mpz_t(), n.get_mpz_t());   // m = c^d mod n
}

compute.cc
#include <gmp.h>    
#include <gmpxx.h>  
#include <iostream>
#include <exception>
#include <fstream>

using namespace std;

void generate_n_bit_long_prime(mpz_class &n, int num_bits) {
    gmp_randstate_t state;
    gmp_randinit_default(state);

    clock_t seed = clock();
    seed = seed / (CLOCKS_PER_SEC / 1000000);
    gmp_randseed_ui(state, seed);
    mpz_urandomb(n.get_mpz_t(), state, num_bits);

    mpz_setbit(n.get_mpz_t(), 0);
    mpz_setbit(n.get_mpz_t(), num_bits - 1);

    while (mpz_probab_prime_p(n.get_mpz_t(), 25) == 0) {
        mpz_add_ui(n.get_mpz_t(), n.get_mpz_t(), 2);
    }

    gmp_randclear(state);
}

mpz_class find_coprime(mpz_class &phi) {
    gmp_randstate_t state;
    gmp_randinit_default(state);

    clock_t seed = clock();
    seed = seed / (CLOCKS_PER_SEC / 1000000);
    gmp_randseed_ui(state, seed);

    mpz_class e;
    e = 65537;

    while (mpz_gcd_ui(NULL, e.get_mpz_t(), phi.get_ui()) != 1) {
        mpz_add_ui(e.get_mpz_t(), e.get_mpz_t(), 2);
    }

    gmp_randclear(state);
    return e;
}

void generate_key_pair(int num_bits, mpz_class &e, mpz_class &d, mpz_class &n) {
    mpz_class p, q, phi;
    generate_n_bit_long_prime(p, num_bits);
    generate_n_bit_long_prime(q, num_bits);
    phi = (p - 1) * (q - 1);
    n = p * q;
    e = find_coprime(phi);
    mpz_invert(d.get_mpz_t(), e.get_mpz_t(), phi.get_mpz_t());
}

void encrypt(mpz_class &m, mpz_class &e, mpz_class &n, mpz_class &c) {
    mpz_powm(c.get_mpz_t(), m.get_mpz_t(), e.get_mpz_t(), n.get_mpz_t());
}

void string_to_number(const std::string& str, mpz_class& n) {
    n = 0;
    for (char c : str) {
        n <<= 8;
        n += c;
    }
}

void number_to_string(const mpz_class& n, std::string& str) {
    str = "";
    mpz_class m = n;
    while (m != 0) {
        char c = m.get_si() & 0xff;
        str = c + str;
        m >>= 8;
    }
}

void decrypt(mpz_class &c, mpz_class &d, mpz_class &n, mpz_class &m) {
    mpz_powm(m.get_mpz_t(), c.get_mpz_t(), d.get_mpz_t(), n.get_mpz_t());
}

int main() {
    cout << "Do you want to encrypt or decrypt? (e/d): ";
    char choice;
    cin >> choice;
    while (choice != 'e' && choice != 'd') {
        cout << "Invalid choice. Please enter e or d: ";
        cin >> choice;
    }
    if (choice == 'e') {
        cout << "Do you want to use a file or standard input? (f/s): ";
        cin >> choice;
        while (choice != 'f' && choice != 's') {
            cout << "Invalid choice. Please enter f or s: ";
            cin >> choice;
        }
        if (choice == 'f') {
            cout << "Enter file name: ";
            string file_name;
            cin >> file_name;
            ifstream file(file_name);
            while (!file.is_open()) {
                cout << "Something is wrong. Please enter a valid file name: ";
                cin >> file_name;
                file.open(file_name);
            }
            string contents;
            string line;
            while (getline(file, line)) {
                contents += line + "\n";
            }
            mpz_class m;
            string_to_number(contents, m);
            cout << "Do you already have a key pair? (y/n): ";
            while (choice != 'y' && choice != 'n') {
                cout << "Invalid choice. Please enter y or n: ";
                cin >> choice;
            }
            if (choice == 'y') {
                cout << "Enter the public key(The e in [n, e]): ";
                mpz_class e;
                cin >> e;
                cout << "Enter the modulus(The n in [n, e]): ";
                mpz_class n;
                cin >> n;
                mpz_class c;
                encrypt(m, e, n, c);
                cout << "Encrypted message: " << c << endl;
                file.close();
            } else if (choice == 'n') {
                mpz_class e, d, n;
                generate_key_pair(4096, e, d, n);
                mpz_class c;
                encrypt(m, e, n, c);
                ofstream output_file("output.txt");
                output_file << "Encrypted message(c): " << c << endl;
                output_file << "Part of private key(d): " << d << endl;
                output_file << "Modulus(n): " << n << endl;
                output_file << "Part of public key(e): " << e << endl;
                output_file.close();
                file.close();
            } else {
                throw bad_exception();
            }
        } else if (choice == 's') {
            cout << "Enter the message: ";
            string message;
            cin.ignore();
            getline(cin, message);
            mpz_class m;
            string_to_number(message, m);
            cout << "Do you already have a key pair? (y/n): ";
            cin >> choice;
            while (choice != 'y' && choice != 'n') {
                cout << "Invalid choice. Please enter y or n: ";
                cin >> choice;
            }
            if (choice == 'y') {
                cout << "Enter the public key(The e in [n, e]): ";
                mpz_class e;
                cin >> e;
                cout << "Enter the modulus(The n in [n, e]): ";
                mpz_class n;
                cin >> n;
                mpz_class c;
                encrypt(m, e, n, c);
                cout << "Encrypted message: " << c << endl;
            } else if (choice == 'n') {
                mpz_class e, d, n;
                generate_key_pair(4096, e, d, n);
                mpz_class c;
                encrypt(m, e, n, c);
                ofstream output_file("output.txt");
                output_file << "Encrypted message(c): " << c << endl;
                output_file << "Part of private key(d): " << d << endl;
                output_file << "Modulus(n): " << n << endl;
                output_file << "Part of public key(e): " << e << endl;
                output_file.close();
            } else {
                throw bad_exception();
            }
        } else {
            throw bad_exception();
        }
    } else if (choice == 'd') {
        cout << "**************************************************";
        cout << "ATTENTION: The file should be in the following format (this is line is not contained in input file):" << endl;
        cout << "Encrypted message(c): " << endl;
        cout << "Part of private key(d): " << endl;
        cout << "Modulus(n): " << endl;
        cout << "**************************************************" << endl;
        cout << "Enter file name: ";
        string file_name;
        cin >> file_name;
        ifstream file(file_name);
        while (!file.is_open()) {
            cout << "Something is wrong. Please enter a valid file name: ";
            cin >> file_name;
            file.open(file_name);
        }
        string line1, line2, line3;
        getline(file, line1);
        getline(file, line2);
        getline(file, line3);
        line1.erase(std::remove_if(line1.begin(), line1.end(), [](char c) { return !std::isdigit(c); }), line1.end());
        line2.erase(std::remove_if(line2.begin(), line2.end(), [](char c) { return !std::isdigit(c); }), line2.end());
        line3.erase(std::remove_if(line3.begin(), line3.end(), [](char c) { return !std::isdigit(c); }), line3.end());
        mpz_class c(line1);
        mpz_class d(line2);
        mpz_class n(line3);
        cout << "c: " << c << endl;
        cout << "d: " << d << endl;
        cout << "n: " << n << endl;
        mpz_class m;
        decrypt(c, d, n, m);
        string message;
        number_to_string(m, message);
        cout << "Decrypted message: " << message << endl;
        file.close();
    } else {
        throw bad_exception();
    }
}
